User管理機能

jaxzzxaj opened the pull request:
#WHAT
User管理機能の作成

＃WHY
ログイン/ログアウトの実装
ログイン状態と未ログイン状態の条件下で変わるユーザーの操作範囲のため

jaxzzxaj committed:
furima用のstylesheetをdrugdrop

jaxzzxaj committed:
furima用のimagesをdrugdrop

jaxzzxaj committed:
productテーブルの作成

・indexアクションへ遷移するページの準備

・先に、ログインページへ遷移するためのindex.htmlを作る。（migrationファイルにdevise用のカラムをコメントアウト）

jaxzzxaj committed:
indexアクション遷移するまでの設定

routes→controllers→viewの設定

jaxzzxaj committed:
レイアウトテンプレートの作成

・_headerと_footerのdrugdrop

・render でsharedファイルの上記を添付させる。

・application.html.erbをfurima素材と置き換える。

jaxzzxaj committed:
/パスの設定

jaxzzxaj committed:
deviseのインストール

gem 'devise’の追加

jaxzzxaj committed:
devise用のuser modelを作成

rails g devise userで追加されたもの。

routes.rbにdevise_forの追加

jaxzzxaj committed:
devise用のuserテーブルの作成

+productテーブルにdeviseのuserとつなぐためにuser_idのカラムを追加

jaxzzxaj committed:
devise用のviewをdrugdrop

jaxzzxaj committed:
テーブルと一致するように編集

tableのカラムとdevise用のnew.html.erbの編集

jaxzzxaj committed:
新規登録、ログインしたときにレコードを取得するように設定

jaxzzxaj committed:
ログインの有無の設定

・ログイン、新規登録、ログアウトのviewのdragdrop

・ <% if user_signed_in? %>でログインの有無

・エラーメッセージの追加

・validationでnullは不可にする。

・userとproductのアソシエーションを結ぶ

・N+1問題の回避

・indexへcurrent.userの名前を表記設定

・ログインしてないひとはログインしてくださいアクションをかける（index内は大丈夫）

chobi555 commented:
こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？

・新規登録に成功する動画
・新規登録に失敗し、エラーが表示される動画
・ログインに成功する動画
・ログインに失敗し、エラーが表示される動画
・ログアウトに成功する動画
・テスト結果の画像

「Gyazo　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

jaxzzxaj committed:
テストコードの準備

・gemの導入後、アプリ内にインストール。

・テスト結果のログを見やすくする設定

・エラーメッセージを英語表記に設定

・テストではない方のエラーメッセージの設定。

jaxzzxaj committed:
テスト用のデーターを設定

gemの導入

・カナ表記＝Gimei
・数字系＝Forgery
・テキスト系＝Faker

＊Factorybotは｛｝必須＊

jaxzzxaj committed:
テストコードの準備

・gemの追加(bundle install,factorybot)
・gemの削除（gimei,forgery）↔factoriesの再修正＋seed.rbの削除
・describe,context,itでテストの準備
・（エラー）app/models/user.rbのclassのend忘れ。

jaxzzxaj committed:
テストコードの記述

・起こったエラー
expectの記述方法は違う
（日本語→英語表記＜ターミナルで書き方を教えてくれる＞）
・expect(@user.errors[:***])とexpect(@user.errors.full.messages)の後の.to includeの記述方法の違い。

jaxzzxaj committed:
ログインの実装忘れ

jaxzzxaj commented:
・新規登録に成功する動画
(gyazo: https://gyazo.com/27d68f6653b26cdd1fd0b70cd4d17714)
・新規登録に失敗し、エラーが表示される動画
(gyazo:https://gyazo.com/4f5e3a199753c811c4b4e6c126754944 )
・ログインに成功する動画
(gyazo: https://gyazo.com/43356f67f4816f6aa6f0b4dd83219f30)
・ログインに失敗し、エラーが表示される動画
(gyazo: https://gyazo.com/10adc5000a66c6577680b85871c0673b)
・ログアウトに成功する動画
(gyazo: https://gyazo.com/35dc75d2f35fd04358e16b0f02fc4e29)
・テスト結果の画像
(gyazo: https://gyazo.com/daee83ee8a6d8d12bab59eb2d289c454 と )https://gyazo.com/ae2e91f01d9a3680a6bbeb66d3ab2fd8

yutamasumoto commented on app/controllers/products_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  before_action :move_to_index, except: :index
+  #下記の記述で@productsの中にすべてのproductテーブルのレコードを取得する。
+  def index
+    @products = Product.includes(:user)
```
```
@products = Product.includes(:user)
```
こちらに関しましては商品一覧表示機能にて実装しますのでコメントアウトしておきましょう。

yutamasumoto commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,11 @@
+  private
+
+  def get_parameters_from_devise
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname,:email,:encrypted_password,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday])
```
```
devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname,:email,:encrypted_password,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday])
```
こちらemailとpasswordに関しましてはdeviseの最初の設定で許可されていますので削除しましょう。

yutamasumoto commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,11 @@
+
+  def get_parameters_from_devise
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname,:email,:encrypted_password,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday])
+    devise_parameter_sanitizer.permit(:sign_in, keys: [:email,:password])
```
```
devise_parameter_sanitizer.permit(:sign_in, keys: [:email,:password])
```
こちらは不要となりますので削除しておきましょう。

yutamasumoto commented on app/controllers/products_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  #下記の記述でnewアクションで記入したuserのすべてのレコード＋ユーザのidを取得。
+  def new
+    @user=User.find(user_params)
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:nickname,:family_name, :second_name, :reader_family_name, :reader_second_name,:birthday).merge(user_id: current_user.id)
+  end
```
こちらuser管理に関連する記述はProductsControllerｍには不要ですので削除しておきましょう。

yutamasumoto commented on app/controllers/products_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    params.require(:user).permit(:nickname,:family_name, :second_name, :reader_family_name, :reader_second_name,:birthday).merge(user_id: current_user.id)
+  end
+
+  def move_to_index
+    unless user_signed_in?
+      redirect_to action: :index
+    end
```
こちらuser管理機能までですと不要な記述なので削除しましょう。

yutamasumoto commented on app/models/product.rb:
```diff
@@ -0,0 +1,14 @@
+class Product < ApplicationRecord
```
Productモデル内の記載はuser管理機能では不要な記述ですのでコメントアウトしておきましょう。

yutamasumoto commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+class User < ApplicationRecord
```
こちらpresence: true以外のバリデーションが設定されていませんのでtrelloの要件を確認しながら必要なバリデーションを追記しましょう。

jaxzzxaj replied:
了解しました。trelloにemailは一意性と書いてあったので、uniqueness: trueは必要と思ったのですがちがうのですね。。

yutamasumoto commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+
+  with_options presence: true do
+    validates :nickname
+    validates :email, uniqueness: true
```
```
validates :email, uniqueness: true
```
こちらはdeviseにて実装されているバリデーションなので削除しましょう。

yutamasumoto commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+  with_options presence: true do
+    validates :nickname
+    validates :email, uniqueness: true
+    validates :encrypted_password
```
```
validates :encrypted_password
```
こちらの:encrypted_passwordに対してのバリデーションは必要かもう一度考えてみましょう。

jaxzzxaj replied:
deviseにはデフォルトでemailとpasswordが設定されているけど確認用のパスワードはされていないと思ったのですが。。

yutamasumoto commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <% render 'shared/error_messages', model: f.object %>
```
```
<% render 'shared/error_messages', model: f.object %>
```
現状ですとエラーメッセージが表示されないと思いますので画面に表示されるように実装しましょう。

jaxzzxaj replied:
申し訳ないです、コミット不足でした！

yutamasumoto commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+      <label class="form-text">パスワード(確認)</label>
+      <span class="indispensable">必須</span>
+    </div>
+    <%= f.password_field :encryted_password, class:"input-default", id:"encryted_password", placeholder:"同じパスワードを入力して下さい" %>
```
```
<%= f.password_field :encryted_password, class:"input-default", id:"encryted_password", placeholder:"同じパスワードを入力して下さい" %>
```
こちらの:encryted_passwordに関しましてはもう一度考えてみましょう。
ヒントといたしましては
deviseを用いてuser管理機能を実装する場合は保存されるカラムはencryted_passwordですが、ビュー側ではpasswordとpassword_confirmationとして実装します。

jaxzzxaj replied:
readmeのときにencrypted_passwordにしたから、viewと違う場合は自分で見つけて編集しよう！という考えなのかと思いました。。

jaxzzxaj replied:
１）controllerでget_parameters_from_deviseでencrypted_passwordを取得してくださいと指示
２）viewのdevise/registration/new.htmlでIdがpassword_confomationなっていたのでencrypted_passwordに変更して取得できるようにする。
３）確認用のパスワード等は空欄はだめですよと条件をつけた。

（作業メモ）
再度trelloを読んでみたらviewの記述内容は変更してはいけないと合ったので、idをpassword_confirmationに戻して、models/user.rbのvalidatesとparamsの取得をpassword_confirmationに変更。マイグレーションファイルのusersテーブルをロールバックしてencrypted_passwordをも同じく変更する。

yutamasumoto commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,105 @@
+      end
+    end
+
+    context "新規登録できない場合" do
```
"テストの検証内容ですが、他にもいろんなパターンを考えてテストを行いましょう。
必要なテストは、タスク一覧の実装の条件にありますので、再確認してみましょう！

yutamasumoto reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

jaxzzxaj committed:
renderの解決とエラー修正

・<%= %>の’＝’忘れ。
・birthdayはstringではなくdate

Katsuno0814 commented on app/controllers/application_controller.rb:
```diff
@@ -1,11 +1,31 @@
   before_action :authenticate_user!, except: :index
-  
+
+  def index
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(users_params)
+    if @user.valid?
+      redirect_tp action: :index
+    else
+      render action: :new
+    end
+  end
```
こちらapplicationcontorllerにメソッドの記述がある場合継承するメソッド全てに
indexやnew、createの内容も引き継がれますので削除いたしましょう。

Katsuno0814 commented on app/controllers/application_controller.rb:
```diff
@@ -1,11 +1,31 @@
   def get_parameters_from_devise
-    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname,:email,:encrypted_password,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday])
-    devise_parameter_sanitizer.permit(:sign_in, keys: [:email,:password])
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname,:email,:password_confirmation,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday])
```
こちらdevise_parameter_sanitizerの中にデフォルトでemailやpassword、password_confirmationは含まれておりますので不要となります！

Katsuno0814 commented on app/controllers/application_controller.rb:
```diff
@@ -1,11 +1,31 @@
+
+  def users_params
+    params.require(:users).permit(:nickname,:email,:encrypted_password,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday)
   end
```
こちらもusers_paramsのメソッドは不要になるため削除お願いいたします。

Katsuno0814 commented on db/migrate/20210402110309_devise_create_users.rb:
```diff
@@ -5,13 +5,13 @@ def change
       ## Database authenticatable
       t.string :email,                                null: false, unique: true, default: ""
-      t.string :encrypted_password,                   null: false, default: ""
+      t.string :password_confirmation,                null: false, default: ""
```
こちらencrypted_passwordのままで問題ありませんのでお手数ではありますが修正お願いいたします。

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いします！

修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

jaxzzxaj committed:
コードの修正

 ・migrationのカラムを戻す
（password_confirmationとは関係ない）

・application.**内は継承されるから二重にしなくて良い（tree）

TakashiKozuma commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
-end
+  before_action :get_parameters_from_devise, if: :devise_controller?
+  before_action :authenticate_user!, except: :index
```
```rb
before_action :authenticate_user!, except: :index
```
こちらの記述は削除し、他の機能を実装する際に、それぞれのコントローラーに記述するようにしましょう。
コントローラーによって```authenticate_user!```を使用したいアクションが異なるためです。

TakashiKozuma commented on app/models/product.rb:
```diff
@@ -0,0 +1,14 @@
+  #テーブル間のアソシエーション（user_idはpresence: trueしなくともよい＝つながっているからデフォルトでなる）
+  belongs_to :user
+
+  with_options presence: true do
+    validates :name
+    validates :category_id
+    validates :shipping_charges_id
+    validates :estimated_shipping_date_id
+    validates :prefecture
+    validates :price
+    validates :text
+  end
```
```rb
  with_options presence: true do
    validates :name
    validates :category_id
    validates :shipping_charges_id
    validates :estimated_shipping_date_id
    validates :prefecture
    validates :price
    validates :text
  end
```
こちらの記述は商品出品機能のものなのでいったんコメントアウトしておきましょう。
Githubのプルリクエストは差分のみが表示される仕様になっています。
そのためこのままだと出品機能を実装した時にレビュワーにこのコードを見せられなくなってしまうためです。

jaxzzxaj replied:
ユーザ管理機能を作るとき、ログインボタンを置ける場所が必要だと思いindex.htmlを作る際に一緒にproductテーブルを作ったのでこのvalidationを作らねばエラーが起きると思ってましたが、いま現段階では使ってないのでエラーが発生してないんですね！了解しました！

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
```
了解しました。trelloにemailは一意性と書いてあったので、uniqueness: trueは必要と思ったのですがちがうのですね。。
```
上記のご返答に対して回答いたします。
下記の記述によりemailの一意制約はdeviseが標準で適用してくれているため、modelへの記述は不要となります。

```rb
  devise :database_authenticatable, :registerable,
         :recoverable, :rememberable, :validatable
```

jaxzzxaj replied:
ああ！この記述はそういう意味だったんですね！

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  has_many :products
+
+  with_options presence: true do
+    validates :nickname
+    validates :password_confirmation
+    validates :family_name
+    validates :second_name
+    validates :reader_family_name
+    validates :reader_second_name
+    validates :birthday
```
```rb
    validates :nickname
    validates :password_confirmation
    validates :family_name
    validates :second_name
    validates :reader_family_name
    validates :reader_second_name
    validates :birthday
```
下記の実装条件を満たすために、バリデーションを追加しましょう。
また、**追加したバリデーションを確認するテストコードも追加しましょう。**
- パスワードは、半角英数字混合での入力が必須であること
- ユーザー本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
- ユーザー本名のフリガナは、全角（カタカナ）での入力が必須であること

jaxzzxaj replied:
（作業メモ）
１）models/user.rbにvalidates :に半角英数字、全角、カタカナのみにする記述の仕方を調べてコードを書いてみる
２）factoriesはそれぞれのカラムに対して一つしか設定してないけど、上記の条件にも当てはまるから記述しなくてもいいと思う。
３）spec/models/user.rbでテストコードの箱？問題用紙？を準備する。
４）（１）で調べたことを使ってテストコードを書いてみる。
５）ターミナルでexample all clearを目指す。

TakashiKozuma commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+<%= render "shared/second-header"%>
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<div class='form-wrap'>
+  <div class='form-header'>
+    <h1 class='form-header-text'>
+      会員情報入力
+    </h1>
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%= render 'shared/error_messages', model: @user %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```rb
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
<div class='form-wrap'>
  <div class='form-header'>
    <h1 class='form-header-text'>
      会員情報入力
    </h1>
  </div>

  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
  <%= render 'shared/error_messages', model: @user %>
  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
3,5,13,15行目のコメントアウトは実装案内なので削除しておきましょう。
実装案内は完成したアプリを人に見せる際に残しておくの望ましくないためです。
他にも「〜しましょう」のような実装案内があれば、実装時に削除するようにしていきましょう。

TakashiKozuma commented on config/routes.rb:
```diff
@@ -1,3 +1,10 @@
+  root to: 'products#index'
+  #テーブルに対して、許可するアクションの設定。
+  resources :products, only: [:index,:new]
+  resources :users, only: [:index,:new,:create]
```
```rb
resources :users, only: [:index,:new,:create]
```
こちらのルーティングは使用しないので削除しましょう。
ユーザーの登録やログイン、ログアウトの機能は2行目の```devise_for :users```で実装されているためです。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
コメントを追記いたしましたので、ご確認お願いします。

jaxzzxaj commented:
下記のテストコードの結果をgyazo

・パスワードは、半角英数字混合での入力が必須であること
・ユーザー本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
・ユーザー本名のフリガナは、全角（カタカナ）での入力が必須であること

URL:　https://gyazo.com/496f78bea43773ab800ec3419a3565d9

jaxzzxaj committed:
コード修正2

validatesの追加とテストコードの追加

・不必要なコードを削除

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+    validates :birthday
+  end
+
+    validates_confirmation_of :email ,format: {with: /\A[a-zA-Z0-9]+\z/, message: "is invalid.Input half-width characters."}
```
```rb
validates_confirmation_of :email ,format: {with: /\A[a-zA-Z0-9]+\z/, message: "is invalid.Input half-width characters."}
```
こちらのバリデーションは英数字混合のものかと存じますので、emailではなくpasswordに適用するように変更しましょう。
実装条件は下記となっています。
- パスワードは、半角英数字混合での入力が必須であること

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,126 @@
+        expect(@user.errors.full_messages).to include("Password is too short (minimum is 6 characters)")
+      end
+
+      it "passwordと確認用passwordの値は一致であること" do
+        @user = build(:user, encrypted_password: '')
+        @user.valid?
+        expect(@user.errors[:password_confirmation]).to include()
+      end
+
+      it "パスワードは、半角英数字混合の入力が必須であること" do
+        @user = build(:user, encrypted_password: '12345')
+        @user.valid?
+        expect(@user.errors[:password_confirmation]).to include()
+      end
```
```ruby
      it "passwordと確認用passwordの値は一致であること" do
        @user = build(:user, encrypted_password: '')
        @user.valid?
        expect(@user.errors[:password_confirmation]).to include()
      end

      it "パスワードは、半角英数字混合の入力が必須であること" do
        @user = build(:user, encrypted_password: '12345')
        @user.valid?
        expect(@user.errors[:password_confirmation]).to include()
      end
```
これらのテストコードはエラー文が出ていないため、異常系のテストができていない状態です。
```@user.password_confirmation = 'passwordと一致しない何か'```や
@user.password = '数字だけ'など記述の仕方を変更し、エラーば運が出るようにしましょう。

jaxzzxaj replied:
新規登録できない場合のerrorの出し方がどうしてもわからないので、新規登録できる場合の方に
"パスワードは、半角英数字混合の入力が必須であること"＋"パスワードは、6文字以上でなければならない"というテストコードを書いたのですが、、、これはアリですか？？

jaxzzxaj replied:
でも、この場合だとアプリケーション内に反映されなくなるのかな。。

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,126 @@
+        @user.valid?
+        expect(@user.errors[:password_confirmation]).to include()
+      end
+
```
passwordに関しては"パスワードは、半角英数字混合での入力が必須であること"を確認するため"半角英字のみでは登録できない"テストと"半角数字のみでは登録できない"の2つのテストを別々に追加しましょう。

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+
+  with_options presence: true do
+    validates :nickname
+    validates :family_name ,format: { with: /\A[ぁ-んァ-ン一-龥々]/, message: "is invalid. Input full-width characters." }
+    validates :second_name ,format: { with: /\A[ぁ-んァ-ン一-龥々]/, message: "is invalid. Input full-width characters." }
```
```rb
    validates :family_name ,format: { with: /\A[ぁ-んァ-ン一-龥々]/, message: "is invalid. Input full-width characters." }
    validates :second_name ,format: { with: /\A[ぁ-んァ-ン一-龥々]/, message: "is invalid. Input full-width characters." }
```
こちらのバリデーションもwith_optionsを用いてまとめましょう。
読み仮名に関するカラムも同様です。
with_optionsの中でwith_optionsを使用することも可能です。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
コメントを追記いたしましたので、ご確認お願いします。

jaxzzxaj committed:
コード修正3

新規登録できない→新規登録できるのほうにテストコードを移した。

kazutaka-div commented on app/models/user.rb:
```diff
@@ -0,0 +1,19 @@
+
+  with_options presence: true do
+    validates :nickname
+    validates :family_name ,format: { with: /\A[ぁ-んァ-ン一-龥々]/, message: "is invalid. Input full-width characters." }
+    validates :second_name ,format: { with: /\A[ぁ-んァ-ン一-龥々]/, message: "is invalid. Input full-width characters." }
+    validates :reader_family_name ,format: { with: /\A[ァ-ヶー－]+\z/, message: "is invalid. Input full-width  katakana characters." }
+    validates :reader_second_name ,format: { with: /\A[ァ-ヶー－]+\z/, message: "is invalid. Input full-width  katakana characters." }
```
バリデーションはwith_optionsを使用することでまとめることができます。
同じ記述は出来るだけ繰り返さないようにしましょう！

kazutaka-div commented on app/models/user.rb:
```diff
@@ -0,0 +1,19 @@
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  has_many :products
```
```ruby
has_many :products
```
上記は、`商品出品機能`の実装になります！
実装時に差分が上がってこないため、現段階ではコメントアウトしましょう！

kazutaka-div commented on db/migrate/20210402110309_devise_create_users.rb:
```diff
@@ -0,0 +1,50 @@
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :email,                                null: false, unique: true, default: ""
```
```ruby
t.string :email,                                null: false, unique: true, default: ""
```
45行目で`unique: true`は定義できているため、上記の`unique: true`は削除しましょう！

jaxzzxaj replied:
ホントだ！ここまで見てなかったです。。

kazutaka-div commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,147 @@
+        expect(@user.errors[:password_confirmation]).to include("doesn't match Password")
+      end
+
+      # it "パスワードは半角英数字混合の入力が必須であること" do
+      #   @user.password ='0000011'
+      #   @user.password ='aaaaabb'
+      #   @user.valid?
+      #   expect(@user.errors[:password_confirmation]).to include()
+      # end
+
+      # it "パスワードが半角英字のみでは登録できない" do
+      #   @user.password = 'aaaaabb'
+      #   @user.valid?
+      #   expect(@user.errors.full_messages).to include()
+      # end
+
+      # it "パスワードが半角数字のみでは登録できない" do
+      #   @user.password = '1111122'
+      #   @user.valid?
+      #   expect(@user.errors.full_messages).to include()
+      # end
```
上記のコメントアウトを外して、テストが通るよう実装しましょう！

jaxzzxaj replied:
validationの記述に誤りがあったみたいでした！

kazutaka-div commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,147 @@
+        expect(@user.errors.full_messages).to include("Birthday can't be blank")
+      end
+
+      it "誕生月が空だと不可" do
+        @user.birthday = ''
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Birthday can't be blank")
+      end
+
+      it "誕生日が空だと不可" do
+        @user.birthday= ''
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Birthday can't be blank")
+      end
```
同じテストが2つあるため、片方を削除しましょう！

jaxzzxaj replied:
誕生年、月、日のテストを書いたつもりでした。userのテーブルには年、月、日すべてひとまとめで＝birthdayとして設定していたので、テストコードもbirthdayで書いてました。このやり方ではだめのでしょうか？

kazutaka-div commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,147 @@
+        expect(@user.errors.full_messages).to include("Reader second name can't be blank")
+      end
+
+      it "ユーザ本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること" do
+        @user.family_name = "kanji"
+        @user.second_name = "hiragana"
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Family name is invalid. Input full-width characters.", "Second name is invalid. Input full-width characters.")
+      end
+
+
+      it "ユーザ本名のフリガナは、全角（カタカナ）での入力が必須であること" do
+        @user.reader_family_name = "漢字"
+        @user.reader_second_name = "ひらがな"
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Reader family name is invalid. Input full-width  katakana characters.", "Reader second name is invalid. Input full-width  katakana characters.")
+      end
```
ユーザー本名を同時にテストするのではなく、片方ずつに分けましょう！
フリガナも同様です！
```ruby
it "ユーザ本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること" do
        @user.family_name = "kanji"
        @user.valid?
        ・・・
end

it "ユーザ本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること" do
        @user.second_name = "hiragana"
        @user.valid?
        ・・・
      end
```

kazutaka-div commented on app/controllers/products_controller.rb:
```diff
@@ -0,0 +1,25 @@
+##class ProductsController < ApplicationController
+  #未ログインuserのみ特定のアクションを制限する設定
+  #before_action :move_to_index, except: :index
+  #下記の記述で@productsの中にすべてのproductテーブルのレコードを取得する。
+  # def index
+  #  @products = Product.includes(:user)
+  # end
+
+  #下記の記述でnewアクションで記入したuserのすべてのレコード＋ユーザのidを取得。
+  # def new
+  #   @user=User.find(user_params)
+  # end
+
+  # private
+
+  # def user_params
+  #   params.require(:user).permit(:nickname,:family_name, :second_name, :reader_family_name, :reader_second_name,:birthday).merge(user_id: current_user.id)
+  # end
+
+  # def move_to_index
+  #   unless user_signed_in?
+  #     redirect_to action: :index
+  #   end
+  # end
+  #end
```
```ruby
##class ProductsController < ApplicationController

#end
```
1行目と25行目のコメントアウトは戻しておきましょう！
現状だと、rails sでアプリが開かない可能性があるためです！
もしアプリが正常に開くようでしたら、こちらは無視していただいて構いません。

jaxzzxaj replied:
コメントアウトを戻したら開くようになりました！

kazutaka-div reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認の上、
再度、レビュー依頼フォームをご提出ください。

jaxzzxaj committed:
コード修正３

validationの半角英数字の記述がおかしかったため、testcodeが引っかかてた。

devise/userを作成するとき一緒にuniqueもついてる（50行周辺を確認）

controllerを置かないとアプリがproduct_controllerが見当たらないよ？とでるから戻す。

jaxzzxaj commented:
申し訳ありません、コードレビューを送信する前にテストコードの結果を添付するのを忘れてしまいました。
もう一度コードレビューを出します！

https://gyazo.com/c6ffbfc502a256d8080c8859427bc8bc

ken-1853 commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,13 @@
+  #deviseではemail,password, password_confirmationはデフォルトで入ってる！
+  def get_parameters_from_devise
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname,:family_name,:second_name,:reader_family_name,:reader_second_name,:birthday])
+    #devise_parameter_sanitizer.permit(:sign_in, keys: [:email,:password])
```
```
 #devise_parameter_sanitizer.permit(:sign_in, keys: [:email,:password])
```

こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめいたします。

他にも不要なコメントアウトが残っていましたら、同様に削除しましょう！

ken-1853 commented on app/views/shared/_header.html.erb:
```diff
@@ -0,0 +1,29 @@
+    </ul>
+    <% if user_signed_in? %>
+    <div class='lists-right'>
+      <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
```
<%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```

こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめいたします

他にも不要なコメントアウトがあれば同様に削除しましょう！

ken-1853 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

細かい指摘になり大変恐縮ですが、よろしくお願い致します！

jaxzzxaj committed:
コメントアウトの削除

ken-1853 commented:
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れさまでした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう。
